eng/pipelines/templates/steps/test.yml

```diff
@@ -9,42 +9,46 @@ steps:
     parameters:
       AgentImage: ${{ parameters.OSName}}
 
+  - template: /eng/common/pipelines/templates/steps/save-package-properties.yml
+    parameters:
+      ServiceDirectory: ${{parameters.ServiceDirectory}}
+
   - script: |
       node common/scripts/install-run-rush.js install
     displayName: "Install dependencies"
 
   - template: /eng/pipelines/templates/steps/set-artifact-packages.yml
     parameters:
-      Artifacts: ${{ parameters.Artifacts }}
+      PackageInfo: $(Build.ArtifactStagingDirectory)/PackageInfo
 
   # Option "-p max" ensures parallelism is set to the number of cores on all platforms, which improves build times.
   # The default on Windows is "cores - 1" (microsoft/rushstack#436).
   - script: |
-      node eng/tools/rush-runner.js build "${{parameters.ServiceDirectory}}" -packages "$(ArtifactPackageNames)" --verbose -p max
+      node eng/tools/rush-runner.js build $(ChangedServices) -packages "$(ArtifactPackageNames)" --verbose -p max
```

Comment:
Also checking here and below, are we basically ALWAYS expecting changed services to be set at this point?

Reply:
> Also checking here and below, are we basically ALWAYS expecting changed services to be set at this point?

If we have called `set-artifact-packages.yml`, YES. That is a safe guarantee.